Mathlib/CategoryTheory/DinatTrans.lean

```diff
@@ -0,0 +1,92 @@
+/-
+Copyright (c) 2023 Andrea Laretto. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrea Laretto
+-/
+import Mathlib.CategoryTheory.Category.Basic
+import Mathlib.CategoryTheory.Functor.Basic
+import Mathlib.CategoryTheory.Opposites
+import Mathlib.CategoryTheory.Products.Basic
+import Mathlib.Tactic.CategoryTheory.Reassoc
+
+/-!
+# Dinatural transformations
+
+Dinatural transformations are special kinds of transformations between
+functors `F G : Cᵒᵖ × C ⥤ D` which depend both covariantly and contravariantly
+on the same category (also known as difunctors).
+
+A dinatural transformation is a family of morphisms given only on *the diagonal* of the two
+functors, and is such that a certain naturality hexagon commutes.
+
+Note that dinatural transformations cannot be composed with each other (since the outer
+hexagon does not commute in general), but can still be "pre/post-composed" with
+ordinary natural transformations.
+-/
+
+namespace CategoryTheory
+
+universe v₁ v₂ v₃ v₄ u₁ u₂ u₃ u₄
+
+variable {C : Type u₁} [Category.{v₁} C] {D : Type u₂} [Category.{v₂} D]
+
+open Opposite
+
+/-- Dinatural transformations between two (di)functors.
+-/
+structure DinatTrans (F G : Cᵒᵖ × C ⥤ D) : Type max u₁ v₂ where
```

Comment:
In `mathlib`, the preferred spelling for "bifunctor" are functors `F G : Cᵒᵖ ⥤ C ⥤ D` to a category of functors. Could you consider using this instead? (This may make easier the formulation of `dinaturality` also.)

Reply:
I tried at first using that, but the actions on objects and morphisms become a bit complicated (I have no idea how to express it without an ad-hoc helper to compose twice to act on the first morphism and then the second), and most importantly stray from the usual category-theoretic notation... I found no other definition for `Bifunctors` other than [this](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/CategoryTheory/Products/Bifunctor.lean), which seems to be using the product definition along with [`hom`](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/CategoryTheory/Functor/Hom.lean).

Reply:
There is little API for "bifunctor" as functors from a product category because there is a focus on functors to a category of functors. I do not see any problem with the following:
```lean
structure DinatTrans (F G : Cᵒᵖ ⥤ C ⥤ D) where
  app (X : C) : (F.obj (op X)).obj X ⟶ (G.obj (op X)).obj X
  dinaturality {X Y : C} (f : X ⟶ Y) :
    (F.map f.op).app X ≫ app X ≫ (G.obj (op X)).map f =
      (F.obj (op Y)).map f ≫ app Y ≫ (G.map f.op).app Y
```
In order to develop the API using this definition, it may become convenient to use some lemmas like `NatTrans.app_naturality` or `NatTrans.naturality_app`.

Reply:
I switched to this representation by defining some helpers in Bifunctor, let me know what you think!